Source/Scene/ModelExperimental/CustomShaderStage.js

```diff
@@ -0,0 +1,171 @@
+    customShader.lightingModel;
+
+  renderResources.uniformMap = combine(
+    renderResources.uniformMap,
+    customShader.uniformMap
+  );
+};
+
+function generateShaderLines(renderResources, customShader, primitive) {
+  var inputLines = {
+    attributeFields: [],
+    vertexInitializationLines: [],
+    fragmentInitializationLines: [],
+  };
+  var attributes = primitive.attributes;
+  for (var i = 0; i < attributes.length; i++) {
+    processAttribute(attributes[i], inputLines);
+  }
+
+  var vertexLines = []
+    .concat(
+      "struct Attributes",
+      "{",
+      inputLines.attributeFields,
+      "};",
+      "",
+      "struct VertexInput",
+      "{",
+      "    Attributes attributes;",
+      "};",
+      "",
+      "void initializeInputStruct(out VertexInput vsInput)",
+      "{",
+      inputLines.vertexInitializationLines,
+      "}"
+    )
+    .join("\n");
+
+  var fragmentLines = []
+    .concat(
+      "struct Attributes",
+      "{",
+      inputLines.attributeFields,
+      "};",
+      "",
+      "struct FragmentInput",
+      "{",
+      "    Attributes attributes;",
+      "    vec3 positionMC;", // model space
+      "    vec3 positionWC;", // World coords (ECEF). Low precision.
+      "    vec3 positionEC;", // Eye coordinates
+      "};",
+      "",
+      "void initializeInputStruct(out FragmentInput fsInput)",
+      "{",
+      "   fsInput.positionMC = v_position;",
+      "   fsInput.positionWC = v_positionWC;",
+      "   fsInput.positionEC = v_positionEC;",
+      inputLines.fragmentInitializationLines,
+      "}"
+    )
+    .join("\n");
+
+  var shaderBuilder = renderResources.shaderBuilder;
+  shaderBuilder.addVertexLines([
+    vertexLines,
+    "#line 0",
+    customShader.vertexShaderText,
+    CustomShaderStageVS,
+  ]);
+  shaderBuilder.addFragmentLines([
+    fragmentLines,
+    "#line 0",
+    customShader.fragmentShaderText,
+    CustomShaderStageFS,
+  ]);
+}
+
+function processAttribute(attribute, inputLines) {
+  var semantic = attribute.semantic;
+  var setIndex = attribute.setIndex;
+  var attributeType = attribute.type;
+
+  var glslType = AttributeType.getGlslType(attributeType);
+
+  var variableName;
+  if (defined(semantic)) {
+    variableName = VertexAttributeSemantic.getVariableName(semantic, setIndex);
+  } else {
+    // Handle user defined vertex attributes. They must begin with an underscore
+    // For example, "_TEMPERATURE" will be converted to "a_temperature".
+    variableName = attribute.name.substring(1).toLowerCase();
+  }
+
+  // Fields for the Attribute struct. for example:
+  // "    vec3 position;"
+  var attributeField = "    " + glslType + " " + variableName + ";";
+
+  // Initializing attribute structs are just a matter of copying the
+  // attribute or varying: E.g.:
+  // "    vsInput.attributes.position = a_position;" -- vertex shader
+  // "    fsInput.attributes.position = v_position;" -- fragment shader
+  var vertexInitialization =
```

Comment:
Oh yikes... default values aren't quite as feasible as I thought... how to determine what _type_ the attributes are? To know all the possible attributes, you'd need to know all the attributes across every primitive of every model of the tileset.

One thing you could do is disable the custom shader for primitives that are not compatible (e.g. their set of attributes doesn't cover the requirements of the shader). What do you think @sanjeetsuhag @lilleyse?

Reply:
wait, actually you could infer the type for built-in things like `normal`, `texCoord_x`, it's only the custom attributes that would be completely unknown. And that could be handled by skipping the primitive with a warning.

Reply:
https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#meshes -- well the inferencing would _nearly_ work, colors can technically be VEC3 _or_ VEC4. I think this would be enough of a corner case that we can just pick one